packages/ds-ext/src/duckDBStorage.ts

```diff
@@ -0,0 +1,97 @@
+import { ObserverStorage } from '@data-story/core';
+import { Database } from 'duckdb-async';
+
+export class DuckDBStorage implements ObserverStorage {
+  private db: Database | null = null;
+  private dbPath: string;
+
+  constructor(dbPath: string) {
+    this.dbPath = dbPath;
+    this.initDatabase();
+  }
+
+  async initDatabase() {
+    this.db = await Database.create(this.dbPath);
+    await this.db.all(`
+      CREATE TABLE IF NOT EXISTS linkCounts (
+        linkId TEXT PRIMARY KEY,
+        count INT,
+        createTime TIMESTAMP,
+        updateTime TIMESTAMP
+      );
+    
+      CREATE TABLE IF NOT EXISTS linkItems (
+        linkId TEXT,
+        item JSON,
+        createTime TIMESTAMP,
+        updateTime TIMESTAMP
+      );
+      CREATE INDEX idx_linkId ON linkItems (linkId);
+    
+      CREATE TABLE IF NOT EXISTS nodes (
+        nodeId TEXT PRIMARY KEY,
+        status TEXT,
+        createTime TIMESTAMP,
+        updateTime TIMESTAMP
+      );
+    `);
+  }
+
+  async close() {
+    if (this.db) {
+      await this.db.close();
+    }
+  }
+
+  async getLinkCount(linkId: string): Promise<number | undefined> {
+    const result = await this.db?.all('SELECT count FROM linkCounts WHERE linkId = ?', linkId);
+    if (result && result.length > 0) {
+      return result[0].count;
+    }
+    return undefined;
+  }
+
+  async setLinkCount(linkId: string, count: number): Promise<void> {
+    const currentTime = new Date();
+    await this.db?.all('INSERT INTO linkCounts (linkId, count, createTime, updateTime) VALUES (?, ?, ?, ?) ON CONFLICT(linkId) DO UPDATE SET count = ?, updateTime = ?',
+      linkId, count, currentTime, currentTime, count, currentTime);
+  }
+
+  async getLinkItems(linkId: string): Promise<Record<string, any>[] | undefined> {
+    const data = await this.db?.all('SELECT item FROM linkItems WHERE linkId = ?', linkId);
+    if (!data || data.length === 0) {
+      return undefined;
+    }
+    const result = data.map(row => JSON.parse(row.item));
+    return result;
+  }
+
+  async setLinkItems(linkId: string, items: Record<string, any>[]): Promise<void> {
+    await this.db?.all('DELETE FROM linkItems WHERE linkId = ?', linkId);
```

Comment:
Is there a reason to prefer DELETE + reinsert over an UPDATE query?

Reply:
During a re-run, I clear all item data corresponding to the `linkId` in the current diagram and insert the item data for the `linkId` from the latest execution run.

If we need to retain the results of each execution going forward, we won't need this reset operation anymore.